PostHog/Replay/URLSessionExtension.swift

```diff
@@ -0,0 +1,195 @@
+            let (data, response) = try await self.data(from: url)
+            let after = Date()
+            captureData(request: nil, response: response, before: before, after: after)
+            return (data, response)
+        } catch {
+            throw error
+        }
+    }
+
+    func postHogUpload(for request: URLRequest, fromFile fileURL: URL) async throws -> (Data, URLResponse) {
+        let before = Date()
+        var after: Date?
+        do {
+            let (data, response) = try await upload(for: request, fromFile: fileURL)
+            after = Date()
+            captureData(request: request, response: response, before: before, after: after)
+            return (data, response)
+        } catch {
+            captureData(request: request, response: nil, before: before, after: after)
+            throw error
+        }
+    }
+
+    func postHogUpload(for request: URLRequest, from bodyData: Data) async throws -> (Data, URLResponse) {
+        let before = Date()
+        var after: Date?
+        do {
+            let (data, response) = try await upload(for: request, from: bodyData)
+            after = Date()
+            captureData(request: request, response: response, before: before, after: after)
+            return (data, response)
+        } catch {
+            captureData(request: request, response: nil, before: before, after: after)
+            throw error
+        }
+    }
+
+    @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+    func postHogData(for request: URLRequest, delegate: (any URLSessionTaskDelegate)? = nil) async throws -> (Data, URLResponse) {
+        let before = Date()
+        var after: Date?
+        do {
+            let (data, response) = try await self.data(for: request, delegate: delegate)
+            after = Date()
+            captureData(request: request, response: response, before: before, after: after)
+            return (data, response)
+        } catch {
+            captureData(request: request, response: nil, before: before, after: after)
+            throw error
+        }
+    }
+
+    @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+    func postHogData(from url: URL, delegate: (any URLSessionTaskDelegate)? = nil) async throws -> (Data, URLResponse) {
+        let before = Date()
+        var after: Date?
+        do {
+            let (data, response) = try await self.data(from: url, delegate: delegate)
+            after = Date()
+            captureData(request: nil, response: response, before: before, after: after)
+            return (data, response)
+        } catch {
+            captureData(request: nil, response: nil, before: before, after: after)
+            throw error
+        }
+    }
+
+    @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+    func postHogUpload(for request: URLRequest, fromFile fileURL: URL, delegate: (any URLSessionTaskDelegate)? = nil) async throws -> (Data, URLResponse) {
+        let before = Date()
```

Comment:
these bodies all seem relatively identical... would they all change at the same time - is it worth extracting "handlers" so it's easier to see what's the same and what's different?

Reply:
The signature has to be identical to the original ones.
I extracted everything to `captureData` except what has to be read as early as possible such as the `before` and `after`.
Since the `captureData` spawns a background thread to capture the network data, there's a bit of duplication, indeed.
What do you mean by `handlers`?

Reply:
so... the content of each method appears to be almost identical

so I'd almost expect (Assuming you can pass functions around in swift)

```

onNetwork(networkFn) {
        let before = Date()
        var after: Date?
        do {
            let (a, b) = networkFn()
            after = Date()
            captureData(request: request, response: response, before: before, after: after)
            return (a, b)
        } catch {
            captureData(request: request, response: nil, before: before, after: after)
            throw error
        }

```

so then you'd have

```
@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
    func postHogDownload(resumeFrom resumeData: Data, delegate: (any URLSessionTaskDelegate)? = nil) async throws -> (URL, URLResponse) {
        onNetwork(() => try await download(resumeFrom: resumeData, delegate: delegate))
    }
```

since they all seem to be methods that wrap a function call with slightly different args and a destructed response 

so when I'm reading the file I can see that all of the swizzled functions do the same thing apart from how they make the network call

but that's coming from a place of basically no swift knowledge so feel free to ignore 😊

Reply:
ah gotcha what you mean

Reply:
done, good idea.